common/logger/src/libra_logger.rs

```diff
@@ -0,0 +1,293 @@
+        self.printer = Some(printer);
+        self
+    }
+
+    pub fn is_async(&mut self, is_async: bool) -> &mut Self {
+        self.is_async = is_async;
+        self
+    }
+
+    pub fn init(&mut self) {
+        self.build()
+    }
+
+    pub fn build(&mut self) {
+        let logger = if self.is_async {
+            let (tx, rx) = mpsc::sync_channel(self.channel_size);
+            let service = LoggerService {
+                rx,
+                // file: None,
+                address: self.address.clone(),
+                printer: self.printer.take(),
+            };
+            let logger = Arc::new(LibraLogger {
+                tx: Some(tx),
+                printer: None,
+                level: self.level,
+            });
+
+            thread::spawn(move || service.run());
+            logger
+        } else {
+            Arc::new(LibraLogger {
+                tx: None,
+                printer: self.printer.take(),
+                level: self.level,
+            })
+        };
+
+        crate::logger::set_global_logger(logger.clone());
+        let logger = Box::leak(Box::new(logger));
+        crate::struct_log::set_struct_logger(logger).expect("Unable to setup structured logger");
+    }
+}
+
+pub struct LibraLogger {
+    tx: Option<SyncSender<StructuredLogEntry>>,
+    printer: Option<Box<dyn Writer>>,
+    level: Level,
+}
+
+impl LibraLogger {
+    pub fn builder() -> LibraLoggerBuilder {
+        LibraLoggerBuilder::new()
+    }
+
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> LibraLoggerBuilder {
+        Self::builder()
+    }
+
+    pub fn init_for_testing() {
+        if env::var("RUST_LOG").is_err() {
+            return;
+        }
+
+        Self::builder()
+            .is_async(false)
+            .printer(Box::new(StderrWriter))
+            .build()
+    }
+
+    fn send_entry(&self, entry: StructuredLogEntry) {
+        if let Some(tx) = &self.tx {
+            if let Err(e) = tx.try_send(entry) {
+                STRUCT_LOG_QUEUE_ERROR_COUNT.inc();
+                eprintln!("Failed to send structured log: {}", e);
+            }
+        } else if let Some(printer) = &self.printer {
+            let s = format(&entry).expect("Unable to format");
+            printer.write(s);
+        }
+    }
+}
+
+impl crate::logger::Logger for LibraLogger {
+    fn enabled(&self, metadata: &crate::Metadata) -> bool {
+        metadata.level() <= self.level
+    }
+
+    fn record(&self, event: &crate::Event) {
+        let mut entry = StructuredLogEntry::default()
+            .level(event.metadata().level())
+            .schemas(event.keys_and_values());
+        entry
+            .add_category(event.metadata().target())
+            .add_module(event.metadata().module_path())
+            .add_location(event.metadata().location());
+
+        if let Some(message) = event.message() {
+            entry = entry.message(message.to_string());
+        }
+
+        self.send_entry(entry)
+    }
```

Comment:
I'm assuming the entry is going to go away as a middleman?

Reply:
Yes, this is just temporary for now